File: cpp/49-GroupAnagrams/main.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>


using namespace std;

class Solution 
{
public:
    vector<vector<string>> groupAnagrams(const vector<string>& strs) 
    {
        unordered_map<string, vector<string>> mappings;

        for(auto key : strs)
        {
            auto original = key;
            sort(key.begin(), key.end());

            mappings[key].push_back(original);
        }

        vector<vector<string>> groups;

        for(auto &pair : mappings)
        {
            groups.push_back(pair.second);
        }

        return groups;
    }
};

int main()
{
    Solution s;
    auto test1 = s.groupAnagrams({"eat","tea","tan","ate","nat","bat"});
}

File: go/49-GroupAnagrams/main.go
package main

import (
	"fmt"
	"sort"
)

func groupAnagrams(strs []string) [][]string {
	mappings := make(map[string][]string)

	for _, key := range strs {
		original := key

		// Sort the characters in the string
		keyEncoding := []byte(key)
		sort.Slice(keyEncoding, func(i, j int) bool { return keyEncoding[i] < keyEncoding[j] })
		sortedKey := string(keyEncoding)

		// And map the sorted key to a sequnece of strings that contain those keys
		mappings[sortedKey] = append(mappings[sortedKey], original)
	}

	groups := make([][]string, 0, len(mappings))

	for _, value := range mappings {
		groups = append(groups, value)
	}

	return groups
}

func main() {
	fmt.Println(groupAnagrams([]string{"eat", "tea", "tan", "ate", "nat", "bat"}))
}
